### PR TITLE
feat: parquet export

### LINKED ISSUE #84: Restructure mongo for Cast
Cast mongo data structures were designed for oryginal version of cast, and then kept and reused.
With current usage they make less sense and could use restructuring.

### PR DESCRIPTION
**Pull request by Rakowskiii**
# Description

Exporting information about user registrations and dispatched messages to data lake.

Resolves #84 

## How Has This Been Tested?

Tests ran, but integration with Data Lake not tested.

## Due Diligence

* [ ] Breaking change
* [ ] Requires a documentation update
* [ ] Requires a e2e/integration test update

### CODE REVIEW

**Comment by Rakowskiii** [comment]
> I think you need to add permissions to the geo-ip dbs bucket in terraform too.

It's already there, just not part of this PR :D
https://github.com/WalletConnect/cast-server/blob/de2e1bcb2f4e4e918fd6d97fd1b07c25e37d55c1/terraform/ecs/iam.tf#L114

**Review by dnul — CHANGES_REQUESTED** [review]
@Rakowskiii let's write into this paths instead:
* s3://walletconnect.data-lake.{env}/cast/messages
* s3://walletconnect.data-lake.{env}/cast/clients

let's use central data-lake bucket `walletconnect.data-lake.{env}`

**Comment by dnul** [comment]
> module.analytics.aws_s3_bucket.analytics-data-lake_bucket must be replaced

the bucket already exists on the data realm so you should just reference it as a variable / local I guess. No need to create a bucket.. 
we can drop `cast` specific analytics buckets since they have no data.

**Comment by dnul** [comment]
@Rakowskiii  I think we might be using an old version of gorgon? 

we've done this integration in the relay-server:

https://github.com/WalletConnect/rs-relay/blob/842d5580a418b5311a4d26eee5cf983b97f32eb0/src/analytics/mod.rs#L60C1-L68C16

they're using this version
gorgon = { git = "https://github.com/WalletConnect/gorgon.git", tag = "v0.5.1" }

sorry to bring all this things now, but we'd rather set data the right way from the beginning.

**Comment by dnul** [comment]
I think this section allows write permissions to the data-lake bucket
https://github.com/WalletConnect/rs-relay/blob/842d5580a418b5311a4d26eee5cf983b97f32eb0/terraform/ecs/analytics.tf#L5

File: terraform/ecs/analytics.tf [hunk 1]
```diff
@@ -0,0 +1,76 @@
+resource "aws_s3_bucket" "analytics_data_lake-bucket" {
+  bucket = var.analytics-data-lake_bucket_name
+}
```
Comment (dnul) [inline terraform/ecs/analytics.tf:3]: @Rakowskiii  still not sure, is this creating a new bucket?  from the output of the terraform diff looks like we're still creating one.

File: terraform/ecs/analytics.tf [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        "Sid" : "ListObjectsInAnalyticsBucket",
+        "Effect" : "Allow",
+        "Action" : ["s3:ListBucket"],
+        "Resource" : ["arn:aws:s3:::${var.analytics-data-lake_bucket_name}"]
+      },
+      {
+        "Sid" : "AllObjectActionsInAnalyticsBucket",
+        "Effect" : "Allow",
+        "Action" : "s3:*Object",
+        "Resource" : ["arn:aws:s3:::${var.analytics-data-lake_bucket_name}/*"]
```
Comment (dnul) [inline terraform/ecs/analytics.tf:25]: can you restrict the path to /cast/* ?

File: Cargo.toml [hunk 3]
```diff
@@ -69,28 +71,33 @@ rand_core = "0.5.0"
...
+once_cell = "1.18.0"
 
 [dev-dependencies]
 test-context = "0.1"
 
 [build-dependencies]
 build-info-build = "0.0"
 
+[features]
+analytics = []
```
Comment (HarryET) [inline Cargo.toml:100]: Feature is nice, is there any case where:
1. Someone runs this themself?
2. Analytics won't be enabled?

If not might be simpler to not include the flag?

File: Dockerfile [hunk 4]
```diff
@@ -52,10 +52,10 @@ RUN                 apt-get update \
 WORKDIR             /app
 # Cache dependancies
 COPY --from=plan    /app/recipe.json recipe.json
-RUN                 RUSTFLAGS="-C link-arg=-fuse-ld=lld" cargo chef cook --recipe-path recipe.json ${RELEASE}
```
Comment (HarryET) [inline Dockerfile:55]: Shouldn't need to use the other linker now

File: Cargo.toml (hunk 3 above)
Comment (Rakowskiii) [reply Cargo.toml]: True, removed it as Cast will only be hosted with them anyways

File: Dockerfile [hunk 5]
```diff
@@ -47,15 +47,15 @@ ENV                 RELEASE=${release:+--release}
...
 # get rid of this.
 RUN                 apt-get update \
-    && apt-get install -y --no-install-recommends protobuf-compiler lld llvm
+    && apt-get install -y --no-install-recommends protobuf-compiler 
 
 WORKDIR             /app
 # Cache dependancies
 COPY --from=plan    /app/recipe.json recipe.json
-RUN                 RUSTFLAGS="-C link-arg=-fuse-ld=lld" cargo chef cook --recipe-path recipe.json ${RELEASE}
+RUN                 cargo chef cook --recipe-path recipe.json ${RELEASE} --analytics analytics
```
Comment (HarryET) [inline Dockerfile:55]: Whats `--analytics analytics`?

Comment (Rakowskiii) [reply Dockerfile]: Lol this was supposed to be feature, must've renamed it by misate
Removed as we got rid of feature and moved analytics into main code

File: src/analytics/mod.rs [hunk 6]
```diff
@@ -43,42 +44,38 @@ impl CastAnalytics {
...
-            )
+            let collector = ParquetWriter::<MessageInfo>::new(opts.clone(), exporter)?;
+            Analytics::new(collector)
         };
 
         let clients = {
-            let exporter = AwsExporter::new(AwsExporterOpts {
-                export_name: "push_clients",
+            let exporter = AwsExporter::new(AwsOpts {
+                export_prefix: "cast/clients",
```
Comment (xav) [inline src/analytics/mod.rs:70]: won't that resolve to `cast/clients/clients`? Feels redundant.

**Review by xav — APPROVED** [review]

File: src/analytics/mod.rs (hunk 6 above)
Comment (Rakowskiii) [reply src/analytics/mod.rs:70]: No, this will properly save them in cast/clients, but the name of the file in `date` folder will be `clients`

Comment (Rakowskiii) [reply src/analytics/mod.rs:70]: image

Comment (Rakowskiii) [reply src/analytics/mod.rs:70]: We looked at it yesterday with @dnul and confirmed this is how we want it to look

Comment (xav) [reply src/analytics/mod.rs:70]: ok then :)

**Review by dnul — APPROVED** [review]
LGTM!

**Pull request merged**